odoo/addons/base/models/ir_async.py

```diff
@@ -0,0 +1,242 @@
+                    "Skipping database %s as its base version is not %s.",
+                    dbname, BASE_VERSION)
+                return
+
+        # process all jobs enqueued in this db
+        while True:
+            with db.cursor() as manager_cr:
+
+                # acquire job
+                manager_cr.execute("""
+                    SELECT *
+                    FROM ir_async
+                    WHERE state = 'created'
+                    LIMIT 1
+                    FOR UPDATE SKIP LOCKED
+                    """)
+                job = manager_cr.dictfetchone()
+                if job is None:
+                    break
+
+                # process job
+                with api.Environment.manage():
+                    with db.cursor() as job_cr:
+                        cls._process_job(job_cr, job)
+
+                # update job
+                manager_cr.execute("""
+                    UPDATE ir_async
+                    SET state=%s,
+                        payload=%s,
+                        write_date=%s
+                    WHERE ID = %s""", (
+                        job['state'],
+                        json.dumps(job['payload']),
+                        datetime.datetime.utcnow(),
+                        job['id'],
+                    ))
+
+    @staticmethod
+    def _process_job(job_cr, job):
+        """
+        Process one job, restore environment and recordset then safely
+        call the desired model method.
+        """
+        for json_field in ('res_ids', 'args', 'kwargs', 'context'):
+            job[json_field] = json.loads(job[json_field])
+        job['context']['async_traceback'] = job['traceback']
+        job['context']['async_job_id'] = job['id']
+        env = api.Environment(job_cr, job['user_id'], job['context'], job['super_user'])
+        ir_async = env['ir.async']
+
+        ir_async._pre_process(job)
+
+        try:
+            _logger.info('Calling "%s" on "%s"', job['method'], job['res_model'])
+            records = env[job['res_model']].browse(job['res_ids'])
+            result = getattr(records, job['method'])(*job['args'], **job['kwargs'])
+            json.dumps(result)  # ensure result is serializable
+            records.flush()
+        except Exception as exc:
+            job['state'] = 'failed'
+            job['payload'] = ir_async._handle_exception(exc)
+        else:
+            _logger.debug("Done, result is %s", result)
+            if result is None:
+                job['state'] = 'done'
+            else:
+                job['state'] = 'succeeded'
+                job['payload'] = ir_async._json_response(result)
+
+        ir_async._post_process(job)
+
+    def _pre_process(self, job):
+        pass
+
+    def _post_process(self, job):
+        pass
+
+    def _merge_traceback(self, tb):
+        """ Merge the given traceback with all previous level of execution """
+        old_tb = self.env.context.get('async_traceback', '').splitlines(keepends=True)
```

Comment:
Passing `job` as a parameter and getting the traceback as `job['traceback']` would be simpler.

Reply:
Héhé, look above :)